### PR TITLE
Remove unhandledRejection

### PR DESCRIPTION
**Pull request by i404788**
This is a follow up to https://github.com/emscripten-core/emscripten/pull/8657

unhandledRejection has been removed, and abort has been adapted to work with that change.

Results of test suite:
```
TEST SUMMARY
    test_core: 4272 run, 4 errors, 14 failures, 392 skipped
    test_sanity: 20 run, 1 errors, 2 failures, 1 skipped
    test_other: 383 run, 3 errors, 2 failures, 31 skipped
```
It seems like the tests that failed are unrelated, but I'll be looking at the results of the CI to see if I missed anything

### CODE REVIEW

**Comment by i404788** [comment]
test_webidl seems to fail I'm unsure of the cause, but I will look into it

**Comment by kripken** [comment]
Why did abort() need to be changed? Seems surprising to me, I guess I'm missing something.

**Comment by i404788** [comment]
@kripken It's a way to remove unhandledRejection without having the behavior change too much. Although it still might be too much change, as you can see from the WebIDL test changes it does change things quite a bit.

I think this is a build up of technical debt, because:
unhandledRejection will break node because on external rejections, so developers would have to make a workaround to patch emscripten's unhandledRejection out of their codebase.
The obvious way to fix that would be to remove unhandledRejection, however unhandledRejection is used to (re-)throw exceptions in `abort` that are async (as exceptions in async context are rejections).
And apparently exiting is desired behavior (https://github.com/emscripten-core/emscripten/pull/5948). That is also seen in test_only_force_stdlibs.

And so a way to fix that is to make sure it exits the first time it hits abort (even if in async context).
A side effect (as you can see in the WebIDL test) is that functions which relied on abort throwing a normal exception will now exit instead. I have yet to a good solution which gives the desired behavior in all cases.

**Comment by i404788** [comment]
Well all required tests pass now, PTAL

**Comment by sbc100** [comment]
Am I right in thinking that this means users won't see a stacktrace in their control on `abort` and/or `assert` anymore?

Also, from what you have written it is advantageous for us not to handle unhandledRejection in general?  (Aside from the cyclic JSON problem)?

**Comment by i404788** [comment]
@sbc100 with ASSERTIONS you will still get a stacktrace, not sure about without.
The main advantage is that developers running in an enviroment where unhandledRejections can happen (e.g. socket handling using libraries) will not have to work around the fact that emscripten is handling rejections. Also if we can confine errors that happen within emscripten code it will also prevent this kind of leaky behavior in the future.

I'm not very confident that this PR is the thing we want to do, as it (while according to tests should be fine) will introduce the new leaky behavior of always exiting in node on abort. Sadly I don't think there is a way to keep the unhandledRejection event to just emscripten code

File: src/postamble.js [hunk 1]
```diff
@@ -419,7 +419,7 @@ function abort(what) {
   EXITSTATUS = 1;
 
 #if ASSERTIONS == 0
-  throw 'abort(' + what + '). Build with -s ASSERTIONS=1 for more info.';
+  console.error('abort(' + what + '). Build with -s ASSERTIONS=1 for more info.');
```
Comment (kripken) [inline src/postamble.js]: these should be calls to `err` instead of `console.error` (as in some environments that is redirected to another place, and `console.*` may not exist).

File: tests/webidl/post.js [hunk 2]
```diff
@@ -1,4 +1,8 @@
 
+process.exit = (e) => {
```
Comment (kripken) [inline tests/webidl/post.js]: probably better to avoid `=>` notation as not all JS engines may support it, and some test environments may run such older VMs.

File: tests/webidl/post.js [hunk 3]
```diff
@@ -1,4 +1,8 @@
...
+  throw e
```
Comment (kripken) [inline tests/webidl/post.js]: but I think the bigger question is, why is this needed in this test?

what worries me is that if we need such a specific hack here, then maybe our users may end up needing hacks of their own.

Comment (i404788) [reply tests/webidl/post.js]: Yeah, that's why I think this PR might not be what we want. try-catch in this way is only used in one test, but I think if we want this behavior it should be it's own test as well.

Comment (kripken) [reply tests/webidl/post.js]: Sorry, I don't follow - is a try-catch in this test a specific problem? which one?

Comment (i404788) [reply tests/webidl/post.js]: try-catch isn't the problem, it's just that it will exit on abort now, so it won't ever reach the other tests. try-catch was supposed to keep it from exiting.
But I will close this PR soon, I've found that the issue most likely lies in undefined symbol macro rather than abort itself.
